### PR TITLE
Native Genode bindings

### PR DESCRIPTION
**Pull request by ehmry**
Bindings for linking Solo5 unikernels into Genode native components. The
bindings are provided as a shared library and may be built as a empty
stub library using the host toolchain or the C++ implementation may be
built using the Genode toolchain and build system. Now that there is a
mix of dynamic and static linked Solo5 implementations the "-static"
flag has been removed from the global LDFLAGS and is now explicitly
specified where required.

---
I am able to build the stub library and tests on Linux and FreeBSD, but I did not test building and linking with the whole Mirage stack on FreeBSD. In the near future I will look into creating small boot image for x86_64 that a Solo5 application can be patched into. If this PR clears then I will make a PR at the main Genode repository for a Solo5 package recipe, and we can build and publish the actual bindings library for Solo5 from code that is hosted here.

Ref https://github.com/genodelabs/genode/issues/2945

### CODE REVIEW

File: tenders/hvt/solo5-hvt-configure [hunk 1]
```diff
@@ -88,7 +88,7 @@ add_module ()
 }
 
 SOLO5_HVT_CFLAGS=
-SOLO5_HVT_LDFLAGS=
+SOLO5_HVT_LDFLAGS=-static
```
Comment (mato) [inline tenders/hvt/solo5-hvt-configure]: Why is this required? `solo5-hvt` is not usually built statically?

File: Makefile.common [hunk 2]
```diff
@@ -34,7 +34,7 @@ MD_CFLAGS+=-mno-red-zone
 endif
 
 # Likewise.
-LDFLAGS=-nostdlib -z max-page-size=0x1000 -static $(HOST_LDFLAGS)
+LDFLAGS=-nostdlib -z max-page-size=0x1000 $(HOST_LDFLAGS)
```
Comment (mato) [inline Makefile.common]: I'm not particularly happy with the way this change is done. Perhaps we could have a `GENODE_LDFLAGS` in here in here instead:

````
COMMON_LDFLAGS=-nostdlib -z max-page-size=0x1000 $(HOST_LDFLAGS)
LDFLAGS=$(COMMON_LDFLAGS) -static
GENODE_LDFLAGS=$(COMMON_LDFLAGS)
````

Then, pass `GENODE_LDFLAGS` in the appropriate places (`.pc` and `%.genode` targets) and keep `LDFLAGS` in all the other places, rather than manually adding `-static`. What do you think?

Comment (ehmry) [reply Makefile.common]: I don't have a problem with that, PR updated.

**Comment by ehmry** [comment]
I've dropped the copywrite and license notice from the linker scripts, those were just boilerplate in the case.

File: solo5-bindings-genode.pc.in [hunk 3]
```diff
@@ -0,0 +1,10 @@
+prefix=${pcfiledir}/../..
+exec_prefix=${prefix}
+includedir=${prefix}/include/solo5-bindings-genode
+libdir=${exec_prefix}/lib/solo5-bindings-genode
+
+Name: solo5-bindings-genode
+Version: 0.4.0
+Description: Solo5 sandboxed execution environment (Genode target)
+Cflags: !CFLAGS! -isystem ${includedir}/crt -I${includedir}/solo5
+Libs: -nostdlib -z max-page-size=0x1000 -gc-sections -Ttext=0x01000000 --dynamic-linker=ld.lib.so --eh-frame-hdr -rpath-link=. -T ${libdir}/genode_dyn.ld ${libdir}/solo5.lib.so
```
Comment (mato) [inline solo5-bindings-genode.pc.in]: This should just be `!GENODE_LDFLAGS!`, shouldn't it? Or is there a 2nd set of Genode LDFLAGS involved here, different from that defined in the Solo5 Makefiles?

File: tests/Makefile.tests [hunk 4]
```diff
@@ -58,6 +61,14 @@ $(BINDINGS_DIR)/muen/solo5_muen.o:
 	$(MAKE) -C $(BINDINGS_DIR) muen
 endif
 
+ifeq ($(BUILD_GENODE), yes)
+%.genode: %.o $(BINDINGS_DIR)/genode/genode_dyn.ld $(BINDINGS_DIR)/genode/solo5_genode.lib.so
+	$(LD) -T $(BINDINGS_DIR)/genode/genode_dyn.ld $(GENODE_LDFLAGS) -o $@ $(BINDINGS_DIR)/genode/solo5.lib.so $< $(LDLIBS)
```
Comment (mato) [inline tests/Makefile.tests]: Isn't this missing `-shared`?

**Comment by mato** [comment]
@ehmry Maybe I'm not reading this right, but it looks to me like you have three separate sets of LDFLAGS for Genode:

1. Those used to build the stub library.
2. Those used to build the solo5 tests (%.genode).
3. Those exported to OPAM via solo5-bindings-genode.pc.

Is this intentional? Can they not all be based off the same (base) values in Makefile.common?

File: tenders/hvt/solo5-hvt-configure (hunk 1 above)
Comment (ehmry) [reply tenders/hvt/solo5-hvt-configure]: Reverted, this is and artifact from the LDFLAGS shuffle.

**Comment by ehmry** [comment]
The LDFLAGS differences are intentional, the library must be linked differently than the applications.

**Comment by ehmry** [comment]
Depends https://github.com/mirage/ocaml-freestanding/pull/43

**Comment by mato** [comment]
@ehmry I merged the ocaml-freestanding dependency, is this ready for another pass of review?

**Comment by ehmry** [comment]
@mato Yes, so far, so good.

**Comment by mato** [comment]
Looking good, I pushed a small fix to `Makefile.tests`, if the CI is green then will merge.
